### PR TITLE
aspects,o/aspectstate: create an aspects API (WIP)

### PR DESCRIPTION
**Pull request by MiguelPires**
This PR adds the `/v2/aspects` endpoint and an `aspectstate` package that expose the current aspects functionality in the API. To test it locally, see `tests/main/aspects/task.yaml` which contains an example interaction with the API. This is a WIP meant to gather early feedback and discuss API design, not for general review

### CODE REVIEW

File: overlord/aspectstate/aspectstate.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+func Set(st *state.State, account, directory, aspect, field, value string) error {
+	st.Lock()
+	defer st.Unlock()
+
+	if err := aspecttest.MaybeMockAspect(st); err != nil {
+		return err
+	}
+
+	var asps map[string]map[string]*aspects.Directory
```
Comment (pedronis) [inline overlord/aspectstate/aspectstate.go:38]: I don't think we want to marshal entire directories, only the data

File: aspects/aspects.go [hunk 2]
```diff
@@ -85,10 +89,52 @@ type Schema interface {
 
 // Directory holds a series of related aspects.
 type Directory struct {
```
Comment (pedronis) [inline aspects/aspects.go:91]: we need to decide when to do the discussed renaming /directory/bundle/

Comment (MiguelPires) [reply aspects/aspects.go:91]: I opened https://github.com/snapcore/snapd/pull/12748. Once it lands in master, I'll merge to this branch so it's up to date

File: overlord/aspectstate/aspectstate.go (hunk 1 above)
Comment (MiguelPires) [reply overlord/aspectstate/aspectstate.go]: Changed it to serialize the databag only and re-create the aspect directory based on the patterns we get from the assertion (which for now is just a list of patterns)

File: aspects/aspects.go
Comment (pedronis) [inline aspects/aspects.go:1]: most of the changes in this file can now be undo now that we only serialize the data bags?

File: overlord/aspectstate/aspectstate.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	if err := asp.Get(field, &value); err != nil {
+		return "", err
+	}
+
+	return value, nil
+}
+
+func updateDatabags(st *state.State, account, directory, aspect string, databag *aspects.JSONDataBag) error {
+	var databags map[string]map[string]map[string]*aspects.JSONDataBag
+	if err := st.Get("databags", &databags); err != nil {
```
Comment (pedronis) [inline overlord/aspectstate/aspectstate.go:113]: aspect-databags?

File: aspects/aspects.go
Comment (MiguelPires) [reply aspects/aspects.go:1]: Yuo're right, I forgot to remove those. Removed them now

File: overlord/aspectstate/aspectstate.go (hunk 3 above)
Comment (MiguelPires) [reply overlord/aspectstate/aspectstate.go]: Changed

File: daemon/api_aspects.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		GET:         getAspect,
+		POST:        setAspect,
+		ReadAccess:  openAccess{},
+		WriteAccess: authenticatedAccess{Polkit: polkitActionManage},
+	}
+)
+
+type AspectRequest struct {
+	Account string `json:"account"`
+	Bundle  string `json:"bundle"`
```
Comment (pedronis) [inline daemon/api_aspects.go:44]: all the place that use bundle string should probably say bundleName for clarity

File: daemon/api_aspects.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	Aspect  string `json:"aspect"`
+	Field   string `json:"field"`
+	Value   string `json:"value"`
```
Comment (pedronis) [inline daemon/api_aspects.go:47]: why Value is a string instead of an interface?

**Review by pedronis — COMMENTED** [review]
thanks, did another pass, noticed a few more things that are slightly off and some comments

**Comment by MiguelPires** [comment]
Closing as this work was split into:
* https://github.com/snapcore/snapd/pull/12824
* https://github.com/snapcore/snapd/pull/12828

**Pull request closed without merging**